File: go/iris-squirrel-example/api/ninja/ninja.go
package ninja

import (
	iris "github.com/kataras/iris/v12"

	"iris-squirrel-example/providers"
	"iris-squirrel-example/types"
)

func RegisterNinja(app *iris.Application, providers *providers.Providers) {
	r := app.Party("/api/ninja")
	{
		// Get ninja
		r.Get("/{id}", func(c iris.Context) {
			id := c.Params().Get("id")
			ninja, err := providers.PGDAL.GetNinja(id)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.JSON(ninja)
		})

		// Insert ninja
		r.Post("/", func(c iris.Context) {
			var ninjaNew types.NinjaNew
			err := c.ReadJSON(&ninjaNew)
			if err != nil {
				c.StatusCode(iris.StatusBadRequest)
				return
			}
			ninja, err := providers.PGDAL.CreateNinja(ninjaNew)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.JSON(ninja)
		})

		// Update ninja
		r.Put("/{id}", func(c iris.Context) {
			id := c.Params().Get("id")
			var ninjaUpdates types.NinjaNew
			err := c.ReadJSON(&ninjaUpdates)
			if err != nil {
				c.StatusCode(iris.StatusBadRequest)
				return
			}
			ninja, err := providers.PGDAL.UpdateNinja(id, ninjaUpdates)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.JSON(ninja)
		})

		// Delete Ninja
		r.Delete("/{id}", func(c iris.Context) {
			id := c.Params().Get("id")
			ninja, err := providers.PGDAL.DeleteNinja(id)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.JSON(ninja)
		})

		// Get ninja with jutsus
		r.Get("/{id}/jutsus", func(c iris.Context) {
			id := c.Params().Get("id")
			ninja, err := providers.PGDAL.GetNinjaWithJutsus(id)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.JSON(ninja)
		})
	}
}

func RegisterNinjaJutsu(app *iris.Application, providers *providers.Providers) {
	r := app.Party("/api/ninja-jutsu")
	{
		// Associate ninja & jutsu
		r.Post("/{ninjaID}/{jutsuID}", func(c iris.Context) {
			ninjaID := c.Params().Get("ninjaID")
			jutsuID := c.Params().Get("jutsuID")
			err := providers.PGDAL.AssociateNinjaJutsu(ninjaID, jutsuID)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.StatusCode(iris.StatusOK)
		})

		// Dissociate ninja & jutsu
		r.Delete("/{ninjaID}/{jutsuID}", func(c iris.Context) {
			ninjaID := c.Params().Get("ninjaID")
			jutsuID := c.Params().Get("jutsuID")
			err := providers.PGDAL.DissociateNinjaJutsu(ninjaID, jutsuID)
			if err != nil {
				c.StatusCode(iris.StatusInternalServerError)
				return
			}
			c.StatusCode(iris.StatusOK)
		})
	}
}

File: py/fastapi-sqlalchemy-example/src/api/ninja.py
"""Ninja API routes."""

# pylint: disable=W0718

from uuid import UUID

from fastapi import APIRouter, FastAPI, Response  # Depends

from ..db.providers import Providers  # get_providers
from ..types.types import NinjaNew, NinjaUpdates


def add_ninja_routes(app: FastAPI, providers: Providers):
    """Add ninja routes."""
    router = APIRouter()

    @router.get("/{ninja_id}")
    async def get_ninja(
        response: Response,
        ninja_id: str,
        # providers: Providers = Depends(get_providers),
    ):
        """Get ninja."""
        try:
            ninja = await providers.pgdal.get_ninja(UUID(ninja_id))
            response.status_code = 200
            return ninja
        except Exception:
            response.status_code = 500
            return {}

    @router.post("/")
    async def create_ninja(
        response: Response,
        ninja_new: NinjaNew,
        # providers: Providers = Depends(get_providers),
    ):
        """Insert new ninja."""
        try:
            ninja = await providers.pgdal.create_ninja(ninja_new)
            response.status_code = 200
            return ninja
        except Exception:
            response.status_code = 500
            return {}

    @router.put("/{ninja_id}")
    async def update_ninja(
        response: Response,
        ninja_id: str,
        ninja_updates: NinjaUpdates,
        # providers: Providers = Depends(get_providers),
    ):
        """Update existing ninja."""
        try:
            ninja = await providers.pgdal.update_ninja(UUID(ninja_id), ninja_updates)
            response.status_code = 200
            return ninja
        except Exception:
            response.status_code = 500
            return {}

    @router.delete("/{ninja_id}")
    async def delete_ninja(
        response: Response,
        ninja_id: str,
        # providers: Providers = Depends(get_providers),
    ):
        """Delete existing ninja."""
        try:
            ninja = await providers.pgdal.delete_ninja(UUID(ninja_id))
            response.status_code = 200
            return ninja
        except Exception:
            response.status_code = 500
            return {}

    @router.post("/{ninja_id}/jutsu/{jutsu_id}")
    async def associate_ninja_and_jutsu(
        response: Response,
        ninja_id: str,
        jutsu_id: str,
        # providers: Providers = Depends(get_providers),
    ):
        """Associate ninja and jutsu."""
        try:
            await providers.pgdal.associate_ninja_and_jutsu(
                UUID(ninja_id), UUID(jutsu_id)
            )
            response.status_code = 200
            return {}
        except Exception:
            response.status_code = 500
            return {}

    @router.delete("/{ninja_id}/jutsu/{jutsu_id}")
    async def dissociate_ninja_and_jutsu(
        response: Response,
        ninja_id: str,
        jutsu_id: str,
        # providers: Providers = Depends(get_providers),
    ):
        """Dissociate ninja and jutsu."""
        try:
            await providers.pgdal.dissociate_ninja_and_jutsu(
                UUID(ninja_id), UUID(jutsu_id)
            )
            response.status_code = 200
            return {}
        except Exception:
            response.status_code = 500
            return {}

    @router.get("/{ninja_id}/jutsus")
    async def get_ninja_with_jutsus(
        response: Response,
        ninja_id: str,
        # providers: Providers = Depends(get_providers),
    ):
        """Get ninja with jutsus."""
        try:
            ninja = await providers.pgdal.get_ninja_with_jutsus(UUID(ninja_id))
            response.status_code = 200
            return ninja
        except Exception:
            response.status_code = 500
            return {}

    app.include_router(
        router,
        prefix="/api/ninja",
    )
